tradeexecutor/state/portfolio.py

```diff
@@ -129,6 +131,57 @@ def get_trade_by_id(self, trade_id: int) -> Optional[TradeExecution]:
     def get_all_positions(self) -> Iterable[TradingPosition]:
         """Get open, closed and frozen, positions."""
         return chain(self.open_positions.values(), self.closed_positions.values(), self.frozen_positions.values())
+    
+    def get_all_positions_filtered(self) -> Iterable[TradingPosition]:
+        """Get open, closed and frozen, positions filtered to remove
+        repaired or failed trades.
+        
+        It also adds bad_data_issues flag to each trade
+        """
+        
+        all_positions = self.get_all_positions()
+        filtered_positions = []
+
+        for position in all_positions:
+            
+            # to avoid copying with same reference
+            filtered_position = copy.deepcopy(position)
+            filtered_position.trades = {}
+            
+            for key, trade in position.trades.items():
+                if trade.is_repaired() or trade.is_repair_trade():
+                    # These trades have quantity set to zero
+                    continue
+
+                # filter out failed trade
+                if trade.executed_at is None:
+                    continue
+                
+                # Internally negative quantities are for sells
+                quantity = trade.executed_quantity
+
+                if trade.planned_mid_price not in (0, None):
+                    price = trade.planned_mid_price
+                else:
+                    # TODO: Legacy trades.
+                    # mid_price is filled to all latest trades
+                    price = trade.executed_price
+
+                    # bad_data_issues only used for remark
+                    trade.bad_data_issues = True
```

Comment:
Get functions should not modify state. 

This part of the code most be redesigned, preferably without any flags